numpy/core/src/multiarray/datetime.c

```diff
@@ -2347,6 +2347,65 @@ get_tzoffset_from_pytzinfo(PyObject *timezone_obj, npy_datetimestruct *dts)
                  get_datetimestruct_minutes(dts));
 }
 
+
+/*
+ * Gets a tzname str by calling tzname() function on
+ * the Python datetime.tzinfo object.
+ */
+NPY_NO_EXPORT char *
+get_tzname_from_pytzinfo(PyObject *timezone_obj, npy_datetimestruct *dts)
+{
+    PyObject *dt, *str;
+
+    /* Create a Python datetime to give to the timezone object */
+    dt = PyDateTime_FromDateAndTime((int)dts->year, dts->month, dts->day,
+                            dts->hour, dts->min, 0, 0);
+    if (dt == NULL) {
+        return "";
+    }
+
+    str = PyObject_CallMethod(timezone_obj, "tzname", "O", dt);
+
+    PyObject *es = PyUnicode_AsEncodedString(str, "utf-8", "~E~");
+    char *s = PyBytes_AS_STRING(es);
+    Py_DECREF(dt);
+    Py_DECREF(str);
+    if (str == NULL) {
+        return "";
+    }
+
+    /* Calculate the tzoffset as the difference between the datetimes */
+    return s;
+}
+
+
+/*
+ * Gets a formatted python datetime str using datetime.strftime().
+ */
+NPY_NO_EXPORT char *
+get_formatted_from_datetimestruct(char *fmt, npy_datetimestruct *dts)
+{
+    PyObject *dt, *str;
+
+    /* Create a Python datetime to give to the timezone object */
+    dt = PyDateTime_FromDateAndTime((int)dts->year, dts->month, dts->day,
+                            dts->hour, dts->min, dts->sec, 0);
+    if (dt == NULL) {
+        return "";
+    }
+
+    str = PyObject_CallMethod(dt, "strftime", "s", fmt);
+
+    PyObject *es = PyUnicode_AsEncodedString(str, "utf-8", "~E~");
+    char *s = PyBytes_AS_STRING(es);
+    Py_DECREF(dt);
+    Py_DECREF(str);
+    if (str == NULL) {
+        return "";
+    }
+    return s;
```

Comment:
I think this leaks a reference to a python strong object

Reply:
Ooops, thanks, I changed to pass the string in as a variable. (same error was in `get_tzname_from_pytzinfo` as well!)

Reply:
I just ran this through a refcount checker, and both tests still leak references:
```
build/testenv/lib/python3.7/site-packages/numpy/core/tests/test_datetime.py::TestDateTime::test_datetime_strftime: leaked references: [6, 6], memory blocks: [6, 6]
build/testenv/lib/python3.7/site-packages/numpy/core/tests/test_datetime.py::TestDateTime::test_datetime_strftime_tz: leaked references: [2, 2], memory blocks: [2, 2]
```

Reply:
OK, may have found another.  I don't know what a refcount checker is. Do you have a link where I can get one?

Reply:
It is fairly easy if you have a debug python installed: https://github.com/numpy/numpy/wiki/Reference-Count-Testing

You also should make sure that you clean up correctly on errors (I realize most of them are not easy to test maybe, but if you can test them, please also test the error paths, that is important, especially in C code). It can be a convenient pattern to just set everything to `NULL` at the definition, and then you can `XDECREF` no matter if there is a failure or things succeeded.

(Have looked a bit, but did not fnid the leak very quickly, I expect it is some simple leak in every call to the formatting function)

Reply:
Its still leaking?  This will be tough for me to debug.  Not sure how easy it is to install a `python3-dbg` on my Mac, and I'm really not very experienced with the python C-API.

Reply:
I can check (I checked with the current last commit before). In any case, it is not a big issue there are likely going to be more reviews coming (maybe I will see too) and that might just flush it out.

Reply:
I ran the leak test on a ubuntu virtual machine.  *All* the tests give the following.  Not sure if I'm doing it correctly....

```
python3-dbg runtests.py -v -- -R 2:3 -s numpy/core/tests/test_datetime.py::TestDateTime::test_datetime_strftime -k "not leaks_references"
```

```
build/testenv/lib/python3.7/site-packages/numpy/core/tests/test_datetime.py::TestDateTime::test_datetime_strftime: leaked references: [77, 77, 77], memory blocks: [23, 23, 23]
build/testenv/lib/python3.7/site-packages/numpy/compat/tests/test_compat.py::test_isfileobj: leaked references: [77, 77, 77], memory blocks: [23, 23, 23]
build/testenv/lib/python3.7/site-packages/numpy/core/tests/test_abc.py::TestABC::test_abstract: leaked references: [77, 77, 77], memory blocks: [23, 23, 23]
```